### PR TITLE
storage: remove PersistMonotonicWriteWorker

### PR DESCRIPTION
**Pull request by danhhz**
It was originally introduced to wrap persist WriteHandles in a task, but
everything that was using it to write data was already in a task. So at
this point, it's just an extra layer of indirection to reason about.

### Motivation

   * This PR refactors existing code.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by danhhz** [comment]
There are some test failures to shake out, but feels like this is a worthwhile simplification?

**Comment by benesch** [comment]
> feels like this is a worthwhile simplification?

1000%!

Ok if I defer the review on this one to @aljoscha? I'm not too familiar with this code these days. (@ParkMyCar may also be able to take a look? I feel like he was the one to add the monotonic write worker originally.)

File: src/storage-controller/src/collection_mgmt.rs [hunk 1]
```diff
@@ -1153,6 +1049,55 @@ where
     (tx, handle.abort_on_drop(), shutdown_tx)
 }
 
+async fn monotonic_append<T: Timestamp + Lattice + Codec64 + TimestampManipulation>(
```
Comment (danhhz) [inline src/storage-controller/src/collection_mgmt.rs:1052]: I'm not entirely sure I ported the intention here correctly, it was kinda hard to follow

File: src/storage-controller/src/lib.rs [hunk 2]
```diff
@@ -718,19 +730,31 @@ where
...
-                DataSource::Introspection(_) => {
+                DataSource::Introspection(typ) => {
                     debug!(data_source = ?collection_state.data_source, meta = ?metadata, "registering {} with persist monotonic worker", id);
-                    self.persist_monotonic_worker.register(id, write);
+                    // We always register the collection with the collection manager,
+                    // regardless of read-only mode. The CollectionManager itself is
+                    // aware of read-only mode and will not attempt to write before told
+                    // to do so.
+                    //
+                    self.register_introspection_collection(id, *typ, write)
```
Comment (danhhz) [inline src/storage-controller/src/lib.rs:740]: Is it safe to pull this call up here? It's hard to tell what in the method is a delicate dance of doing things in _just_ the right order and what's historical cruft

**Review by ParkMyCar — APPROVED** [review]
LGTM from the Monotonic writer perspective, not sure about the 0dt and read-only bits though

File: src/storage-controller/src/collection_mgmt.rs [hunk 3]
```diff
@@ -1153,6 +1049,55 @@ where
...
+        let updates = updates
+            .iter()
+            .map(|TimestamplessUpdate { row, diff }| {
+                ((SourceData(Ok(row.clone())), ()), lower.clone(), diff)
+            })
+            .collect::<Vec<_>>();
+        let res = write_handle
+            .compare_and_append(
+                updates,
+                Antichain::from_elem(lower),
```
Comment (aljoscha) [inline src/storage-controller/src/collection_mgmt.rs:1086]: This is where it differs from the original impl, which uses `current_upper` as the expected upper and only uses `lower` to determine a timestamp for the timestampless updates.

As is here, I would expect this to fail when `at_least` is higher than `current_upper`, because the `compare_and_append` would then fail because the expected upper doesn't line up.

File: src/storage-controller/src/lib.rs [hunk 4]
```diff
@@ -1630,11 +1636,21 @@ where
...
-            .persist_monotonic_worker
-            .recent_upper(id)
+        let metadata = &self.storage_collections.collection_metadata(id)?;
+        let persist_client = self
+            .persist
+            .open(metadata.persist_location.clone())
             .await
-            .expect("sender hung up")?;
+            .unwrap();
+        let write_handle = self
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/storage-controller/src/lib.rs:1645]: Write handles are cheap as long as you never run a CaA with them! That's subtle, so probably worth a comment. I'll stick one in when I circle back to this.

File: src/storage-controller/src/lib.rs [hunk 5]
```diff
@@ -718,19 +730,31 @@ where
...
                     debug!(data_source = ?collection_state.data_source, meta = ?metadata, "registering {} with persist monotonic worker", id);
-                    self.persist_monotonic_worker.register(id, write);
                     self.collections.insert(id, collection_state);
                     new_source_statistic_entries.insert(id);
                     // This collection of statistics is periodically aggregated into
                     // `source_statistics`.
                     new_webhook_statistic_entries.insert(id);
+                    // Register the collection so our manager knows about it.
+                    //
+                    // NOTE: Maybe this shouldn't be in the collection manager,
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/storage-controller/src/lib.rs:753]: :joy:

File: src/storage-controller/src/collection_mgmt.rs (hunk 3 above)
Comment (danhhz) [reply src/storage-controller/src/collection_mgmt.rs:1086]: Interesting! I'll have to load this up again to feel like I totally understand that, but sounds like a pretty mechanical fix.

Something I don't understand is that previously, the `MonotonicAppend` call could return a `StorageError` that the upper didn't match. But why? Isn't the contract just that the thing gets written down at some timestamp that's at least the `at_least`? Do you know if this was just an artifact of implementation (assigning a timestamp and then afterward treating it like a normal Append) or is that necessary for some reason?

File: src/storage-controller/src/lib.rs [hunk 6]
```diff
@@ -869,9 +893,6 @@ where
             }
         }
 
-        self.append_shard_mappings(new_collections.into_iter(), 1)
```
Comment (danhhz) [inline src/storage-controller/src/lib.rs:872]: AFAICT, this call was a no-op before because it would hit this short-circuit:

```
        let id = match self
            .introspection_ids
            .lock()
            .expect("poisoned")
            .get(&IntrospectionType::ShardMapping)
        {
            Some(id) => *id,
            _ => return,
        };
```

Comment (danhhz) [reply src/storage-controller/src/lib.rs:872]: Oh no, this was incorrect. It was a no-op for the first call, but not on the later ones

**Comment by danhhz** [comment]
Okay @aljoscha this seems to be passing tests now. Modulo a lint failure and the comment I promised, this should be ready for a review!

File: src/storage-controller/src/collection_mgmt.rs (hunk 3 above)
Comment (aljoscha) [reply src/storage-controller/src/collection_mgmt.rs]: I think the intention was "we pick a timestamp that's guaranteed to increase, but we still do a compare_and_append to ensure that there's no other writer". I also believe that the reasoning around these "monotonic appends" is a bit hand-wave-y, especially when you consider potential concurrent writers, in a use-case isolation future, and just in general. It comes from a world where we didn't have the differential collections and from a world where we didn't think about 0dt/uci and relied a lot on "yeah yeah, `envd` fencing makes this all correct" (which I never fully trusted... 😅 ).

So it isn't what I would call a _blind write_, where the semantics really are: write down the thing, no matter what, if there's a concurrent write and the upper doesn't match you can retry.

It's part of the reason I introduced the differential collections, which have a sync-and-retry loop and generally feel much more robust. And now that the "partially truncate" logic also used a proper compare_and_append (I changed that in [1]), I feel better about our consistency story. And it might be that all users of `monotonic_append` now are okay with them being blind writes.

[1]:
https://github.com/MaterializeInc/materialize/blob/f5b681abad68572d46b650a11f5196ae3d39eea4/src/storage-controller/src/lib.rs#L3343

File: src/storage-controller/src/lib.rs [hunk 7]
```diff
@@ -2982,19 +3018,21 @@ where
...
         introspection_type: IntrospectionType,
+        recent_upper: Antichain<T>,
+        write_handle: Option<&mut WriteHandle<SourceData, (), T, Diff>>,
     ) -> Result<(), StorageError<T>> {
         tracing::info!(%id, ?introspection_type, "preparing introspection collection for writes");
 
         match introspection_type {
             IntrospectionType::ShardMapping => {
-                self.initialize_shard_mapping().await;
+                // Done by the `self.append_shard_mappings` call.
```
Comment (aljoscha) [inline src/storage-controller/src/lib.rs:3028]: Don't we still need this? The `append_shard_mappings` call in `create_collections` only does it for new collections, and `initialize_shard_mapping()` made sure we write down the collections that were already known at the time at which we "create" the shard-mappings shard.

**Review by aljoscha — COMMENTED** [review]
Overall, this is excellent! 🙌 

I had one inline concern about `initialize_shard_mapping()`.

File: src/storage-controller/src/lib.rs (hunk 7 above)
Comment (danhhz) [reply src/storage-controller/src/lib.rs:3028]: Yeah, not sure how we want this to work. I removed this because they were getting double registered. It seems that before, during initial boot we were hitting the following short-circuit when `append_shard_mappings` got called as part of `create_collections_for_bootstrap`. But this PR shifts the order slightly of things in `create_collections_for_bootstrap` and ends up putting `ShardMapping` in the `introspection_ids` map before `append_shard_mappings` gets called. This resulted in them getting added twice, once from `append_shard_mappings` and once from `initialize_shard_mapping`

```
        let id = match self
            .introspection_ids
            .lock()
            .expect("poisoned")
            .get(&IntrospectionType::ShardMapping)
        {
            Some(id) => *id,
            _ => return,
        };
```

Comment (danhhz) [reply src/storage-controller/src/lib.rs:3040]: (We hashed this out in a huddle. Decision was to make this short-circuit a panic if that's possible, or add a comment if not.)

**Comment by danhhz** [comment]
@aljoscha Nightlies seem happy with replacing the append_shard_mapping short-circuit with a panic. Turns out it was already documented to do that, anyway. Wanna stamp?

**Comment by aljoscha** [comment]
Sorry, should have stamped yesterday after we talked! 🙈

**Review by aljoscha — APPROVED** [review]

**Comment by danhhz** [comment]
\o/ TFTR!
